internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestLeaderReplicaFailover.java

```diff
@@ -0,0 +1,228 @@
+        .setHybridRewindSeconds(5)
+        .setHybridOffsetLagThreshold(2)
+        .setNativeReplicationEnabled(true);
+    createStoreForJob(clusterName, keySchemaStr, valueSchemaStr, props, storeParms).close();
+    // Create a new version
+    VersionCreationResponse versionCreationResponse;
+    versionCreationResponse = TestUtils.assertCommand(
+        parentControllerClient.requestTopicForWrites(
+            storeName,
+            10 * 1024,
+            Version.PushType.BATCH,
+            Version.guidBasedDummyPushId(),
+            true,
+            true,
+            false,
+            Optional.empty(),
+            Optional.empty(),
+            Optional.empty(),
+            false,
+            -1));
+
+    String topic = versionCreationResponse.getKafkaTopic();
+    String kafkaUrl = versionCreationResponse.getKafkaBootstrapServers();
+    VeniceWriterFactory veniceWriterFactory = TestUtils.getVeniceWriterFactory(kafkaUrl);
+    HelixAdmin admin = new ZKHelixAdmin(clusterWrapper.getZk().getAddress());
+    try (VeniceWriter<byte[], byte[], byte[]> veniceWriter = veniceWriterFactory.createBasicVeniceWriter(topic)) {
+      veniceWriter.broadcastStartOfPush(true, Collections.emptyMap());
+      Map<byte[], byte[]> sortedInputRecords = generateData(100, true, 0, serializer);
+      for (Map.Entry<byte[], byte[]> entry: sortedInputRecords.entrySet()) {
+        veniceWriter.put(entry.getKey(), entry.getValue(), 1, null);
+      }
+      veniceWriter.broadcastEndOfPush(Collections.emptyMap());
+    }
+    // Wait push completed.
+    TestUtils.waitForNonDeterministicCompletion(60, TimeUnit.SECONDS, () -> {
+      try {
+        return clusterWrapper.getLeaderVeniceController()
+            .getVeniceAdmin()
+            .getOffLinePushStatus(clusterWrapper.getClusterName(), topic)
+            .getExecutionStatus()
+            .equals(ExecutionStatus.COMPLETED);
+      } catch (Exception e) {
+        return false;
+      }
+    });
+
+    HelixBaseRoutingRepository routingDataRepo = clusterWrapper.getLeaderVeniceController()
+        .getVeniceHelixAdmin()
+        .getHelixVeniceClusterResources(clusterName)
+        .getRoutingDataRepository();
+
+    Instance leader = routingDataRepo.getLeaderInstance(topic, 0);
+
+    // Wait until all partitions have ready-to-serve instances
+    TestUtils.waitForNonDeterministicAssertion(20, TimeUnit.SECONDS, () -> {
+      RoutingDataRepository routingDataRepository = clusterWrapper.getRandomVeniceRouter().getRoutingDataRepository();
+      Assert.assertTrue(routingDataRepository.containsKafkaTopic(topic));
+      int partitionCount = routingDataRepository.getPartitionAssignments(topic).getAllPartitions().size();
+      for (int partition = 0; partition < partitionCount; partition++) {
+        Assert.assertTrue(routingDataRepository.getReadyToServeInstances(topic, partition).size() > 2);
```

Comment:
Can we verify there would be some instances, which have disabled replicas and the ready-to-serve instances should be 2 instead of 3?

Reply:
Sure added the checks.